### PR TITLE
Apply changes from remote database to sqlproj - sql-database-projects changes

### LINKED ISSUE #13308: Adding feature to schema compare database/dacpac to project, and update project with changes
**Is your feature request related to a problem? Please describe.**
Schema compare and SQL projects go hand-in-hand to keep a good overview over changes, and makes deploying these changes to the database easy. The problem is when there are changes that have been made on the database that have not come from the project.

**Describe the solution or feature you'd like**
Would it be possible to add the feature to schema compare from a database to a project, and then update the project with the changes?

**Describe alternatives you've considered**
This is a feature in Visual Studio, but i would like to use Azure Data Studio for this workflow, but the option is not there.

**Additional context**
The alternative for the user is to either script the changed/added object manually and add it to the project, or having to recreate the entire project every time to make sure all changes were included.

### PR DESCRIPTION
**Pull request by Benjin**
Restarting this PR for fresh eyes, fresh comments

----

The last of the main reviews for updating SQL projects from a database, which fixes #13308.  This set of changes are mostly in sql-database-projects, though there's also some stitch-up in schema-compare

Also fixes https://github.com/microsoft/azuredatastudio/issues/17782 and fixes https://github.com/microsoft/azuredatastudio/issues/17788

New dialog:
image

Button with icon in database dashboard:
image

### CODE REVIEW

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 1]
```diff
@@ -1244,6 +1324,125 @@ export class ProjectsController {
...
+				hexValues[tmp >> 20 & 0xF] +
+				hexValues[tmp >> 24 & 0xF] +
+				hexValues[tmp >> 28 & 0xF];
+		}
+
+		// 'Set the two most significant bits (bits 6 and 7) of the clock_seq_hi_and_reserved to zero and one, respectively'
+		let clockSequenceHi: string = hexValues[8 + (Math.random() * 4) | 0];
+		return oct.substr(0, 8) + '-' + oct.substr(9, 4) + '-4' + oct.substr(13, 3) + '-' + clockSequenceHi + oct.substr(16, 3) + '-' + oct.substr(19, 12);
+		/* tslint:enable:no-bitwise */
+	}
```
Comment (Benjin) [inline extensions/sql-database-projects/src/controllers/projectController.ts:1444]: Replace with package

File: extensions/sql-database-projects/src/test/dialogs/updateProjectFromDatabaseDialog.test.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+		const profile = mockConnectionProfile;
+		const dialog = new UpdateProjectFromDatabaseDialog(profile, project);
+		await dialog.openDialog();
+		await dialog.populatedInputsPromise;
+
+		should.equal((<any>dialog.serverDropdown!.value).displayName, profile.options['connectionName'], `Server dropdown should be "${profile.options['connectionName']}", but instead was "${(<any>dialog.serverDropdown!.value).displayName}".`);
+		should.equal(dialog.databaseDropdown!.value, profile.databaseName, `Database dropdown should as "${profile.databaseName}", but instead was "${dialog.databaseDropdown!.value}".`);
+		should.equal(dialog.projectFileTextBox!.value, project.projectFilePath, `Project file textbox should be the sqlproj path (${project.projectFilePath}), but instead was "${dialog.projectFileTextBox!.value}".`);
+
```
Comment (Benjin) [inline extensions/sql-database-projects/src/test/dialogs/updateProjectFromDatabaseDialog.test.ts:68]: remove blank line

File: extensions/sql-database-projects/package.json [hunk 3]
```diff
@@ -252,6 +270,14 @@
         {
           "command": "sqlDatabaseProjects.createProjectFromDatabase"
         },
+        {
+          "command": "sqlDatabaseProjects.updateProjectFromDatabase",
+          "when": "false"
+        },
+        {
+          "command": "sqlDatabaseProjects.updateProjectFromDatabase",
```
Comment (Benjin) [inline extensions/sql-database-projects/package.json:278]: should be updateDsp

File: extensions/sql-database-projects/src/common/constants.ts [hunk 4]
```diff
@@ -240,9 +241,27 @@ export const selectProjectLocation = localize('selectProjectLocation', "Select p
 export const ProjectParentDirectoryNotExistError = (location: string): string => { return localize('dataworkspace.projectParentDirectoryNotExistError', "The selected project location '{0}' does not exist or is not a directory.", location); };
 export const ProjectDirectoryAlreadyExistError = (projectName: string, location: string): string => { return localize('dataworkspace.projectDirectoryAlreadyExistError', "There is already a directory named '{0}' in the selected location: '{1}'.", projectName, location); };
 
+// Update Project From Database dialog strings
+
+export const updateProjectFromDatabaseDialogName = localize('updateProjectFromDatabaseDialogName', "Update project from database");
+export const updateProjectDialogOkButtonText = localize('updateProjectDialogOkButtonText', "Update");
+export const noSqlProjFile = localize('noSqlProjFile', "The selected project file does not exist");
+export const noSchemaCompareExtension = localize('noSchemaCompareExtension', "The schema-compare extension needs to be downloaded to a update a project from a database");
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/constants.ts:249]: "Schema Compare"

File: extensions/sql-database-projects/src/test/dialogs/updateProjectFromDatabaseDialog.test.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		await dialog.openDialog();
+		await dialog.populatedInputsPromise;
+
+		should.equal((<any>dialog.serverDropdown!.value).displayName, profile.options['connectionName'], `Server dropdown should be "${profile.options['connectionName']}", but instead was "${(<any>dialog.serverDropdown!.value).displayName}".`);
+		should.equal(dialog.databaseDropdown!.value, profile.databaseName, `Database dropdown should be "${profile.databaseName}", but instead was "${dialog.databaseDropdown!.value}".`);
+		should.equal(dialog.projectFileTextBox!.value, '', `Project file textbox should not be populated, but instead was "${dialog.projectFileTextBox!.value}".`);
+		should.equal(dialog.dialog.okButton.enabled, false, 'Okay button should be disabled.');
+	});
+
+	it.only('Should populate endpoints correctly when context is complete', async function (): Promise<void> {
```
Comment (Benjin) [inline extensions/sql-database-projects/src/test/dialogs/updateProjectFromDatabaseDialog.test.ts:55]: remove `only`

File: extensions/schema-compare/src/schemaCompareMainWindow.ts [hunk 6]
```diff
@@ -844,7 +844,12 @@ export class SchemaCompareMainWindow {
 					case mssql.SchemaCompareEndpointType.Database:
 						result = await service.schemaComparePublishDatabaseChanges(this.comparisonResult.operationId, this.targetEndpointInfo.serverName, this.targetEndpointInfo.databaseName, azdata.TaskExecutionMode.execute);
 						break;
-					case mssql.SchemaCompareEndpointType.Project: // Project apply needs sql-database-projects updates in (circular dependency; coming next) // TODO: re-add this and show project logic below
+					case mssql.SchemaCompareEndpointType.Project: // Project apply needs sql-database-projects updates in (circular dependency; coming next)
+						result = await vscode.commands.executeCommand(loc.sqlDatabaseProjectsPublishChanges, this.comparisonResult.operationId, this.targetEndpointInfo.projectFilePath, this.targetEndpointInfo.folderStructure);
+						if (!result.success) {
+							await vscode.window.showErrorMessage(loc.applyError);
```
Comment (kisantia) [inline extensions/schema-compare/src/schemaCompareMainWindow.ts:850]: this can probably be void instead of await

File: extensions/schema-compare/src/schemaCompareMainWindow.ts [hunk 7]
```diff
@@ -870,6 +875,11 @@ export class SchemaCompareMainWindow {
 						'endTime': Date.now().toString(),
 						'operationId': this.comparisonResult.operationId
 					}).send();
+
+				if (this.targetEndpointInfo.endpointType === mssql.SchemaCompareEndpointType.Project) {
+					vscode.commands.executeCommand(loc.sqlDatabaseProjectsShowProjectsView);
+					await vscode.window.showInformationMessage(loc.applySuccess);
```
Comment (kisantia) [inline extensions/schema-compare/src/schemaCompareMainWindow.ts:881]: can also be void unless we need to wait for the user to dismiss this

File: extensions/sql-database-projects/package.json [hunk 8]
```diff
@@ -148,11 +150,22 @@
         "category": "%sqlDatabaseProjects.displayName%",
         "icon": "images/databaseProjectToolbar.svg"
       },
+      {
+        "command": "sqlDatabaseProjects.updateProjectFromDatabase",
+        "title": "%sqlDatabaseProjects.updateProjectFromDatabase%",
+        "category": "%sqlDatabaseProjects.displayName%",
+        "icon": "images/databaseProjectToolbar.svg"
```
Comment (kisantia) [inline extensions/sql-database-projects/package.json:157]: can you add a screenshot where this icon is used?

File: extensions/sql-database-projects/package.nls.json [hunk 9]
```diff
@@ -9,6 +9,8 @@
 	"sqlDatabaseProjects.build": "Build",
 	"sqlDatabaseProjects.publish": "Publish",
 	"sqlDatabaseProjects.createProjectFromDatabase": "Create Project From Database",
+	"sqlDatabaseProjects.updateProjectFromDatabase": "Update Project From Database",
+	"sqlDatabaseProjects.schemaCompareGetDsp": "Get Data Schema Provider",
```
Comment (kisantia) [inline extensions/sql-database-projects/package.nls.json:13]: database instead of data

File: extensions/sql-database-projects/package.json [hunk 10]
```diff
@@ -186,6 +199,11 @@
           "when": "view == dataworkspace.views.main",
           "group": "1_currentWorkspace@1"
         },
+        {
+          "command": "sqlDatabaseProjects.updateProjectFromDatabase",
+          "when": "view == dataworkspace.views.main",
```
Comment (kisantia) [inline extensions/sql-database-projects/package.json:204]: is there a default project when launched from command palette? What if there aren't any projects open in the workspace?

File: extensions/sql-database-projects/src/common/constants.ts [hunk 11]
```diff
@@ -240,9 +241,27 @@ export const selectProjectLocation = localize('selectProjectLocation', "Select p
 export const ProjectParentDirectoryNotExistError = (location: string): string => { return localize('dataworkspace.projectParentDirectoryNotExistError', "The selected project location '{0}' does not exist or is not a directory.", location); };
 export const ProjectDirectoryAlreadyExistError = (projectName: string, location: string): string => { return localize('dataworkspace.projectDirectoryAlreadyExistError', "There is already a directory named '{0}' in the selected location: '{1}'.", projectName, location); };
 
+// Update Project From Database dialog strings
+
+export const updateProjectFromDatabaseDialogName = localize('updateProjectFromDatabaseDialogName', "Update project from database");
+export const updateProjectDialogOkButtonText = localize('updateProjectDialogOkButtonText', "Update");
```
Comment (kisantia) [inline extensions/sql-database-projects/src/common/constants.ts:247]: the const name can just be UpdateText so it's easier to reuse somewhere else in the future

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 12]
```diff
@@ -1244,6 +1324,125 @@ export class ProjectsController {
...
+
+		if (model.action === UpdateAction.Update) {
+			const publishResult = await this.schemaComparePublishProjectChanges(
+				operationId, model.targetEndpointInfo.projectFilePath, model.targetEndpointInfo.folderStructure
+			);
+
+			if (publishResult.success) {
+				await vscode.window.showInformationMessage(constants.applySuccess);
+			} else {
+				await vscode.window.showErrorMessage(constants.applyError);
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:1407]: it would be good to also show the error

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 13]
```diff
@@ -453,23 +455,29 @@ export class ProjectsController {
...
+					sourceParam = source as azdataType.IConnectionProfile;
+				}
+
+				try {
+					TelemetryReporter.sendActionEvent(TelemetryViews.ProjectController, TelemetryActions.projectSchemaCompareCommandInvoked);
+					await vscode.commands.executeCommand(constants.schemaCompareStartCommand, sourceParam, targetParam, undefined);
+				} catch (e) {
 					throw new Error(constants.buildFailedCannotStartSchemaCompare);
 				}
+
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:476]: extra blank lines

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 14]
```diff
@@ -486,6 +494,78 @@ export class ProjectsController {
...
 
+	public async schemaCompareGetTargetScripts(projectFilePath: string): Promise<string[]> {
+		const project = await Project.openProject(projectFilePath);
+		let files: string[] = [];
+
+		const fileEntries: FileProjectEntry[] = project.files;
+
+		fileEntries.forEach(f => {
+
+			if (f.fsUri.fsPath.endsWith('.sql')) {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:505]: use sql const

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 15]
```diff
@@ -486,6 +494,78 @@ export class ProjectsController {
...
+				files.push(f.fsUri.fsPath);
+			}
+		});
+
+		return files;
+	}
+
+	public async schemaCompareGetDsp(projectFilePath: string): Promise<string> {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:513]: this doesn't look like it's schema compare specific. Maybe just rename it to getProjectDsp?

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 16]
```diff
@@ -486,6 +494,78 @@ export class ProjectsController {
...
+		return project.getProjectTargetVersion();
+	}
+
+	public async schemaComparePublishProjectChanges(operationId: string, projectFilePath: string, folderStructure: string): Promise<mssql.SchemaComparePublishProjectResult> {
+		let ext = vscode.extensions.getExtension(mssql.extension.name)!;
+		const service = (await ext.activate() as mssql.IExtension).schemaCompare;
+
+		const projectPath = path.dirname(projectFilePath);
+
+		let fs: mssql.ExtractTarget = mssql.ExtractTarget.flat;
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:524]: why set it it to flat here when it'll get set again in the switch?

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 17]
```diff
@@ -486,6 +494,78 @@ export class ProjectsController {
...
+		toRemove.forEach(f => toRemoveEntries.push(new FileProjectEntry(f, f.path.replace(projectPath + '\\', ''), EntryType.File)));
+
+		toRemoveEntries.forEach(async f => await project.exclude(f));
+
+		await this.buildProject(project);
+
+		return result;
+	}
+
+	public async schemaCompareShowProjectsView() {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:564]: this isn't schema compare specific. Other places in this file also use `workspaceApi.showProjectsView()`, so maybe add a helper function for this that can also be used in those places

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts [hunk 18]
```diff
@@ -0,0 +1,610 @@
...
+
+import { Deferred } from '../common/promise';
+import { Project } from '../models/project';
+import { cssStyles } from '../common/uiConstants';
+import { IconPathHelper } from '../common/iconHelper';
+import { UpdateDataModel, UpdateAction } from '../models/api/update';
+import { exists, getAzdataApi, getDataWorkspaceExtensionApi } from '../common/utils';
+import * as path from 'path';
+
+export class UpdateProjectFromDatabaseDialog {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:23]: please add screenshots for this new dialog

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts [hunk 19]
```diff
@@ -0,0 +1,610 @@
...
+				});
+
+			let formModel = this.formBuilder.component();
+			await view.initializeModel(formModel);
+			this.connectionButton?.focus();
+			this.initDialogPromise.resolve();
+		});
+	}
+
+	private createServerRow(view: azdata.ModelView): azdata.FlexContainer {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:150]: looks like this is a lot of the same components that are built in the other dialogs. Doesn't have to be in this PR, but it would be good to have a base dialog class or something so that the different dialogs don't each have to implement creating these common UI components

File: extensions/schema-compare/src/schemaCompareMainWindow.ts (hunk 6 above)
Comment (Charles-Gagnon) [reply extensions/schema-compare/src/schemaCompareMainWindow.ts:850]: Specially it *should* be since otherwise you'll be waiting here forever until the user dismisses the message (same for the other ones pointed out)

Comment (Benjin) [reply extensions/schema-compare/src/schemaCompareMainWindow.ts:850]: Good catch

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts (hunk 19 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:150]: Agreed.  Within an extension is easy enough, but thinking more broadly: what are our options for sharing code between extensions?

File: extensions/sql-database-projects/package.json (hunk 8 above)
Comment (Benjin) [reply extensions/sql-database-projects/package.json:157]: Added (database dashboard icon)

File: extensions/sql-database-projects/package.json (hunk 10 above)
Comment (Benjin) [reply extensions/sql-database-projects/package.json:204]: This command is disabled for the command palette.  Having said that, it'd be completely fine to launch from there; the dialog can seeded with any combination (or lack) of database and project.

File: extensions/sql-database-projects/package.json (hunk 8 above)
Comment (kisantia) [reply extensions/sql-database-projects/package.json:157]: the toolbar is starting to look a little crowded with the long names of "Create project from database" and "Update project from database." How much do we think people will use this from the server/database toolbar?

**Comment by kisantia** [comment]
It looks like the Action radio buttons aren't lined up correctly with the label. The label should be lined up with the first radio button, like in the schema compare dialog

File: extensions/sql-database-projects/package.json [hunk 20]
```diff
@@ -252,6 +270,14 @@
         {
           "command": "sqlDatabaseProjects.createProjectFromDatabase"
         },
+        {
+          "command": "sqlDatabaseProjects.updateProjectFromDatabase",
+          "when": "false"
+        },
+        {
+          "command": "sqlDatabaseProjects.schemaCompareGetDsp",
```
Comment (kisantia) [inline extensions/sql-database-projects/package.json:278]: should this be `sqlDatabaseProjects.getProjectDatabaseSchemaProvider`?

File: extensions/sql-database-projects/src/controllers/mainController.ts [hunk 21]
```diff
@@ -56,6 +57,10 @@ export default class MainController implements vscode.Disposable {
 		vscode.commands.registerCommand('sqlDatabaseProjects.build', async (node: WorkspaceTreeItem) => { return this.projectsController.buildProject(node); });
 		vscode.commands.registerCommand('sqlDatabaseProjects.publish', async (node: WorkspaceTreeItem) => { return this.projectsController.publishProject(node); });
 		vscode.commands.registerCommand('sqlDatabaseProjects.schemaCompare', async (node: WorkspaceTreeItem) => { return this.projectsController.schemaCompare(node); });
+		vscode.commands.registerCommand('sqlDatabaseProjects.schemaCompareGetTargetScripts', async (projectFilePath: string): Promise<string[]> => { return await this.projectsController.schemaCompareGetTargetScripts(projectFilePath); });
+		vscode.commands.registerCommand('sqlDatabaseProjects.getProjectDatabaseSchemaProvider', async (projectFilePath: string): Promise<string> => { return await this.projectsController.getProjectDatabaseSchemaProvider(projectFilePath); });
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/mainController.ts:61]: I think it would make more sense to expose these in projectProvider.ts rather than have them as commands since they're being used directly by another extension, not used through command palette or a right-click menu by the user.

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 22]
```diff
@@ -1244,6 +1321,100 @@ export class ProjectsController {
 		// TODO: Check for success; throw error
 	}
 
+	public async updateProjectFromDatabase(context: azdataType.IConnectionProfile | mssqlVscode.ITreeNodeInfo | dataworkspace.WorkspaceTreeItem): Promise<UpdateProjectFromDatabaseDialog> {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:1324]: add doc comments

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 23]
```diff
@@ -1244,6 +1321,100 @@ export class ProjectsController {
...
+
+	public getUpdateProjectFromDatabaseDialog(connection: azdataType.IConnectionProfile | mssqlVscode.IConnectionInfo | undefined, project: Project | undefined): UpdateProjectFromDatabaseDialog {
+		return new UpdateProjectFromDatabaseDialog(connection, project);
+	}
+
+	public async updateProjectFromDatabaseCallback(model: UpdateDataModel) {
+		try {
+			await this.updateProjectFromDatabaseApiCall(model);
+		} catch (err) {
+			await vscode.window.showErrorMessage(utils.getErrorMessage(err));
```
Comment (kisantia) [inline extensions/sql-database-projects/src/controllers/projectController.ts:1357]: void instead of await

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 12 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/controllers/projectController.ts:1404]: also should be void instead of await here and the info message above

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts (hunk 19 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:150]: I vaguely remember @Charles-Gagnon has looked into this a bit in the past?

Comment (Charles-Gagnon) [reply extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:150]: *shrug* A small bit but we don't really have many options besides just doing an npm package (we might be able to use the same ts file in multiple extensions, but I've never seen that done so not sure how easy it'd be to set up). 

But I'd like to see what specific cases we have for sharing UI components like that across extensions that aren't something we would be adding to core instead before we start really digging into that. So if you have any cases let's chat and see what we can come up with.

File: extensions/sql-database-projects/src/controllers/mainController.ts (hunk 21 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/controllers/mainController.ts:61]: I think you're right.  Think that's something that could be done after this gets checked in so that it's guaranteed to meet askmode?  I think it'd be okay if that changeover ends up happening for Feb due to holiday rush and my OoF.

Comment (kisantia) [reply extensions/sql-database-projects/src/controllers/mainController.ts:61]: I'd prefer it get done right before checking in, but sure it's ok since we're tight on time (as long as this actually gets fixed soon).

File: extensions/sql-database-projects/package.json [hunk 24]
```diff
@@ -293,6 +319,11 @@
           "when": "view == dataworkspace.views.main && viewItem == databaseProject.itemType.project && azdataAvailable",
           "group": "1_dbProjectsFirst@3"
         },
+        {
+          "command": "sqlDatabaseProjects.updateProjectFromDatabase",
+          "when": "view == dataworkspace.views.main && viewItem == databaseProject.itemType.project",
```
Comment (kisantia) [inline extensions/sql-database-projects/package.json:324]: do we need to hide this command in the vscode extension because there's no schema compare in vscode?

File: extensions/sql-database-projects/src/controllers/mainController.ts (hunk 21 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/controllers/mainController.ts]: Aight, I've swapped the commands for ProjectProvider functions.

File: extensions/sql-database-projects/src/projectProvider/projectProvider.ts [hunk 25]
```diff
@@ -168,4 +168,19 @@ export class SqlDatabaseProjectProvider implements dataworkspace.IProjectProvide
 		const projectUri = getDataWorkspaceExtensionApi().openSpecificProjectNewProjectDialog(projectType);
 		return projectUri;
 	}
+
+	/**
+ * Gets the list of .sql scripts contained in a project
+ * @param projectFilePath
```
Comment (kisantia) [inline extensions/sql-database-projects/src/projectProvider/projectProvider.ts:174]: indent

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts [hunk 26]
```diff
@@ -0,0 +1,610 @@
...
+			this.populateServerDropdown();
+		}
+	}
+
+	private createProjectLocationRow(view: azdata.ModelView): azdata.FlexContainer {
+		const browseFolderButton: azdata.Component = this.createBrowseFileButton(view);
+
+		const value = this.project ? this.project.projectFilePath : '';
+
+		this.projectFileTextBox = view.modelBuilder.inputBox().withProps({
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:367]: has having a dropdown with the open projects(like the add database reference dialog) instead of an input box with browse button been considered?

File: extensions/sql-database-projects/src/projectProvider/projectProvider.ts (hunk 25 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/projectProvider/projectProvider.ts:174]: Huh, wonder why the linter didn't catch this

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts (hunk 26 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:367]: Don't know that it has.  I wouldn't mind swapping to a dropdown with "Browse..." at the end because I think there's some value to being able to target an arbitrary csproj, but maybe that change to a dropdown can happen in a different check-in?

File: extensions/schema-compare/src/dialogs/schemaCompareDialog.ts [hunk 27]
```diff
@@ -141,8 +142,8 @@ export class SchemaCompareDialog {
 			this.schemaCompareMainWindow.sourceEndpointInfo = {
 				endpointType: mssql.SchemaCompareEndpointType.Project,
 				projectFilePath: this.sourceTextBox.value,
-				targetScripts: await this.getTargetScripts(true),
-				dataSchemaProvider: await this.getDsp(this.sourceTextBox.value),
+				targetScripts: await this.getProjectScriptFiles(this.sourceTextBox.value),
+				dataSchemaProvider: await this.getDatabaseSchemaProvider(this.sourceTextBox.value),
```
[the remark this answers is not in the document]
Comment (Benjin) [reply extensions/schema-compare/src/dialogs/schemaCompareDialog.ts]: I agree, though this requires an update on the Tools Service side first.  I'll make that change in another wave of PRs

File: extensions/schema-compare/src/schemaCompareMainWindow.ts [hunk 28]
```diff
@@ -159,14 +163,38 @@ export class SchemaCompareMainWindow {
...
+	/**
+	 * Primary functional entrypoint for opening the schema comparison window, and optionally running it.
+	 * @param source
+	 * @param target
+	 * @param runComparison whether to immediately run the schema comparison.  Requires both source and target to be specified.  Cannot be true when comparisonResult is set.
+	 * @param comparisonResult a pre-computed schema comparison result to display.  Cannot be set when runComparison is true.
+	 */
+	public async launch(source: mssql.SchemaCompareEndpointInfo | undefined, target: mssql.SchemaCompareEndpointInfo | undefined, runComparison: boolean = false, comparisonResult: mssql.SchemaCompareResult | undefined) {
+		if (runComparison && comparisonResult) {
+			throw new Error('Cannot both pass a comparison result and request a new comparison be run.');
```
Comment (kisantia) [inline extensions/schema-compare/src/schemaCompareMainWindow.ts:178]: Do these error messages not need to be localized because they won't be shown to the user?

File: extensions/schema-compare/src/schemaCompareMainWindow.ts [hunk 29]
```diff
@@ -159,14 +163,38 @@ export class SchemaCompareMainWindow {
...
+	}
+
+	/**
+	 * Primary functional entrypoint for opening the schema comparison window, and optionally running it.
+	 * @param source
+	 * @param target
+	 * @param runComparison whether to immediately run the schema comparison.  Requires both source and target to be specified.  Cannot be true when comparisonResult is set.
+	 * @param comparisonResult a pre-computed schema comparison result to display.  Cannot be set when runComparison is true.
+	 */
+	public async launch(source: mssql.SchemaCompareEndpointInfo | undefined, target: mssql.SchemaCompareEndpointInfo | undefined, runComparison: boolean = false, comparisonResult: mssql.SchemaCompareResult | undefined) {
```
Comment (kisantia) [inline extensions/schema-compare/src/schemaCompareMainWindow.ts:176]: add return type of function

File: extensions/sql-database-projects/package.json [hunk 30]
```diff
@@ -254,6 +267,14 @@
         {
           "command": "sqlDatabaseProjects.createProjectFromDatabase"
         },
+        {
+          "command": "sqlDatabaseProjects.updateProjectFromDatabase",
+          "when": "false"
+        },
+        {
+          "command": "sqlDatabaseProjects.getProjectDatabaseSchemaProvider",
```
Comment (kisantia) [inline extensions/sql-database-projects/package.json:275]: this command isn't needed anymore now that it got added to sqldbproj.IExtension, right?

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts [hunk 31]
```diff
@@ -0,0 +1,610 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+/* eslint-disable @typescript-eslint/no-floating-promises */
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:6]: why was this added? We usually try to avoid disabling the lint rules

File: extensions/schema-compare/src/schemaCompareMainWindow.ts (hunk 28 above)
Comment (Benjin) [reply extensions/schema-compare/src/schemaCompareMainWindow.ts:178]: Correct.

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts (hunk 31 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts]: Looks like something leftover from the initial PR.  Removed

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts [hunk 32]
```diff
@@ -142,7 +140,7 @@ export class UpdateProjectFromDatabaseDialog {
 
 			let formModel = this.formBuilder.component();
 			await view.initializeModel(formModel);
-			this.connectionButton?.focus();
+			await this.connectionButton?.focus();
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:143]: this isn't necessary unless this isn't the first focusable element. Alan made a change so that the focus is automatically on the first focusable element in dialogs when they're opened

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts (hunk 32 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/dialogs/updateProjectFromDatabaseDialog.ts:143]: The first focusable element is the textbox for the server name; we want it to be the button for selecting a connection.

**Comment by m4ttheweric** [comment]
@Benjin Any word on when we can find this in either the main release or insiders build? Reallllllly need this : )

**Comment by kisantia** [comment]
@m4ttheweric we're targeting getting this out in insiders the week of Feb 14th

**Comment by kisantia** [comment]
@m4ttheweric update project is now available in insiders with schema compare (1.13.0) and sql database projects (0.15.0)!
